File: Python/maximum-depth-of-binary-tree.py
#Time Complexity:   O(n)
#Space Complexity:  O(n)
#Speed: 98.14%
#Memory: 50.93%

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        else:
            return max(self.maxDepth(root.left)+1, self.maxDepth(root.right)+1)

File: C++/Maximum-Depth-of-Binary-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        
        //breadth first search
        
        if (root == nullptr) {
            
            return 0;
            
        }
        
        int count = 0;
        queue<TreeNode*> treeQueue;
        treeQueue.push(root);
        
        while (!treeQueue.empty()) {
            
            count++;
            
            for (int i = 0; i < treeQueue.size(); i++) {
                
                TreeNode* frontNode = treeQueue.front();
                treeQueue.pop();
                
                if (frontNode->left != nullptr) {
                    
                    treeQueue.push(frontNode->left);
                    
                    
                }
                
                if (frontNode->right != nullptr) {
                    
                    treeQueue.push(frontNode->right);
                                        
                }
                              
                
                
            }
                      
            
        }
        
        return count;
        
    }
};